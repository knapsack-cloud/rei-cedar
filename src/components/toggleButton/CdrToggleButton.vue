<template>
  <ul :class="style['segmented-control']" role="radiogroup" aria-label="things-foo">
    <li
      v-for="(item, index) in items"
      :key="`${item}-${index}`"
      :class="style['segmented-control__item']"
      role="presentation"
    >
      <button 
        role="radio"
        :ref="el => { toggleButtonElements[index] = el }"
        :aria-checked="selectedIndex === index ? true : false"
        :tabindex="focusedIndex === index ? 0 : -1"
        @click.prevent="selectToggleButton($event, index)"
        @keyup.right="focusNext(index)"
        @keyup.left="focusPrev(index)"
      >
        {{ item }}
      </button>
    </li>
  </ul>
</template>

<script setup>
import { ref, provide, useSlots, onMounted, nextTick, computed, useCssModule } from 'vue';
const props = defineProps({
  items: {
    required: true,
    type: Array,
    validator: (value) => {
      const result = value.every((item) => {
        if (typeof item !== 'string') {
          console.error('All toggleButton items must be of type string', item);
          return false;
        }
        return true;
      });
      return result;
    },
  },
});
const selectedIndex = ref(0);
const focusedIndex = ref(0);
const toggleButtonElements = ref([]);
const style = useCssModule();

const selectToggleButton = (event, index) => {
  selectedIndex.value = index;
  focusedIndex.value = index;
  toggleButtonElements.value[index].focus();
}

const focusNext = (index) => {
  // const isLastButton = (focusedIndex.value === toggleButtonElements.value.length - 1);
  // if (isLastButton) {
  //   return;
  // }

  const nextIndex = index + 1;
  focusedIndex.value = nextIndex;
  toggleButtonElements.value[nextIndex].focus();
}

const focusPrev = (index) => {
  // const isLastButton = (focusedIndex.value === toggleButtonElements.value.length - 1);
  // if (isLastButton) {
  //   return;
  // }

  const prevIndex = index - 1;
  focusedIndex.value = prevIndex;
  toggleButtonElements.value[prevIndex].focus();
}

</script>


<style lang="scss" module src="./styles/CdrToggleButton.module.scss">
</style>
