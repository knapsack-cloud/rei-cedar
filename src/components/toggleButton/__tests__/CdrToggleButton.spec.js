import { afterEach, beforeEach, describe, expect, it } from 'vitest';
import { shallowMount, mount } from '../../../../test/vue-jest-style-workaround.js';
import CdrToggleButton from '../CdrToggleButton.vue';

describe('CdrToggleButton', () => {
  let wrapper;
  let firstButton;
  let secondButton;
  let thirdButton;
  describe('component snapshot', () => {
    beforeEach(()=>{
      wrapper = mount(CdrToggleButton, {
        propsData: {
          items: ['thing-1', 'thing-2', 'thing-3']
        },
        attachTo: document.body
      });
      firstButton = wrapper.findAll('button')[0];
      secondButton = wrapper.findAll('button')[1];
      thirdButton = wrapper.findAll('button')[2];
      firstButton.element.focus();
    })

    it('matches the snapshot', () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    it('the first button is focused',  () => {
      expect(document.activeElement).toBe(firstButton.wrapperElement);
    });

    it('the tabindex of the first button is 0', ()=>{
      expect(firstButton.attributes('tabindex')).toBe('0');
    })

    it('the tabindex of the second button is -1', ()=>{
      expect(secondButton.attributes('tabindex')).toBe('-1');
    });

    it('the aria-checked value of the first button is true', ()=>{
      expect(firstButton.attributes('aria-checked')).toBe('true');
    })

    it('the aria-checked value of the second button is false', ()=>{
      expect(secondButton.attributes('aria-checked')).toBe('false');
    });

    describe('when hitting the right arrow key', ()=>{
      beforeEach(()=>{
        firstButton.trigger('keyup.right');
      });
      afterEach(()=>{
        firstButton.element.focus();
      });

      it('the second button is focused', ()=> {
        expect(document.activeElement).toBe(secondButton.wrapperElement);
      })

      it('the second button has a tabindex of 0', ()=> {
        expect(secondButton.attributes('tabindex')).toBe('0');
      })
    })

    describe('when hitting the right arrow key', ()=>{
      beforeEach(async ()=>{
        thirdButton.element.focus();
        await wrapper.vm.$nextTick();
        thirdButton.trigger('keyup.left');
      });
      
      afterEach(()=>{
        thirdButton.element.focus();
      });

      it('the second button is focused', ()=> {
        expect(document.activeElement).toBe(secondButton.wrapperElement);
      })

      it('the second button has a tabindex of 0', ()=> {
        expect(secondButton.attributes('tabindex')).toBe('0');
      })
    });

    describe('when the second button is selected and the right arrow key has been hit', ()=>{
      beforeEach(()=>{
        secondButton.trigger('click');
        secondButton.trigger('keyup.right');
      });

      it('the second button has a tabindex of 0', ()=>{
        expect(secondButton.attributes('tabindex')).toBe('0');
      })

      // it('the third button has a tabindex of 0', ()=>{
      //   expect(thirdButton.attributes('tabindex')).toBe('0');
      // })

    })

    //Test for when the third button is selected and you're hitting right

    describe('when clicking on an unchecked button', ()=>{
      beforeEach(()=>{
        secondButton.trigger('click');
      });

      it('matches snapshot', () => {
        expect(wrapper.element).toMatchSnapshot();
      });

      it('turns the aria-checked value to true', () => {
        expect(secondButton.attributes('aria-checked')).toBe('true');
      });

      it('turns the aria-checked value of the previously checked button to false', () => {
        expect(firstButton.attributes('aria-checked')).toBe('false');
      });

      it('the tabindex of the first button is -1', ()=>{
        expect(firstButton.attributes('tabindex')).toBe('-1');
      });

      it('the tabindex of the second button is 0', ()=>{
        expect(secondButton.attributes('tabindex')).toBe('0');
      });

      it('the second button is focused', ()=> {
        expect(document.activeElement).toBe(secondButton.wrapperElement);
      });
    });
  })
});
