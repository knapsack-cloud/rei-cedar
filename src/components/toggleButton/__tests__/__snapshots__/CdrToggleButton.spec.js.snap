// Vitest Snapshot v1

exports[`CdrToggleButton > component snapshot > adding the full-width prop > matches the snapshot 1`] = `
<ul
  aria-label="things-food"
  class="cdr-toggle-group cdr-toggle-group--full-width cdr-toggle-group--medium"
  role="radiogroup"
>
  
  <li
    class="cdr-toggle-button__container"
    role="presentation"
  >
    <button
      aria-checked="true"
      class="cdr-toggle-button__item"
      role="radio"
      tabindex="0"
      value="thing-1"
    >
      
      thing-1
      
    </button>
  </li>
  <li
    class="cdr-toggle-button__container"
    role="presentation"
  >
    <button
      aria-checked="false"
      class="cdr-toggle-button__item"
      role="radio"
      tabindex="-1"
      value="thing-2"
    >
      
      thing-2
      
    </button>
  </li>
  <li
    class="cdr-toggle-button__container"
    role="presentation"
  >
    <button
      aria-checked="false"
      class="cdr-toggle-button__item"
      role="radio"
      tabindex="-1"
      value="thing-3"
    >
      
      thing-3
      
    </button>
  </li>
  
</ul>
`;

exports[`CdrToggleButton > component snapshot > adding the large prop > matches the snapshot 1`] = `
<ul
  aria-label="things-food"
  class="cdr-toggle-group cdr-toggle-group--large"
  role="radiogroup"
>
  
  <li
    class="cdr-toggle-button__container"
    role="presentation"
  >
    <button
      aria-checked="true"
      class="cdr-toggle-button__item"
      role="radio"
      tabindex="0"
      value="thing-1"
    >
      
      thing-1
      
    </button>
  </li>
  <li
    class="cdr-toggle-button__container"
    role="presentation"
  >
    <button
      aria-checked="false"
      class="cdr-toggle-button__item"
      role="radio"
      tabindex="-1"
      value="thing-2"
    >
      
      thing-2
      
    </button>
  </li>
  <li
    class="cdr-toggle-button__container"
    role="presentation"
  >
    <button
      aria-checked="false"
      class="cdr-toggle-button__item"
      role="radio"
      tabindex="-1"
      value="thing-3"
    >
      
      thing-3
      
    </button>
  </li>
  
</ul>
`;

exports[`CdrToggleButton > component snapshot > matches the snapshot 1`] = `
<ul
  aria-label="things-food"
  class="cdr-toggle-group cdr-toggle-group--medium"
  role="radiogroup"
>
  
  <li
    class="cdr-toggle-button__container"
    role="presentation"
  >
    <button
      aria-checked="true"
      class="cdr-toggle-button__item"
      role="radio"
      tabindex="0"
      value="thing-1"
    >
      
      thing-1
      
    </button>
  </li>
  <li
    class="cdr-toggle-button__container"
    role="presentation"
  >
    <button
      aria-checked="false"
      class="cdr-toggle-button__item"
      role="radio"
      tabindex="-1"
      value="thing-2"
    >
      
      thing-2
      
    </button>
  </li>
  <li
    class="cdr-toggle-button__container"
    role="presentation"
  >
    <button
      aria-checked="false"
      class="cdr-toggle-button__item"
      role="radio"
      tabindex="-1"
      value="thing-3"
    >
      
      thing-3
      
    </button>
  </li>
  
</ul>
`;

exports[`CdrToggleButton > component snapshot > when clicking on an unchecked button > matches snapshot 1`] = `
<ul
  aria-label="things-food"
  class="cdr-toggle-group cdr-toggle-group--medium"
  role="radiogroup"
>
  
  <li
    class="cdr-toggle-button__container"
    role="presentation"
  >
    <button
      aria-checked="false"
      class="cdr-toggle-button__item"
      role="radio"
      tabindex="-1"
      value="thing-1"
    >
      
      thing-1
      
    </button>
  </li>
  <li
    class="cdr-toggle-button__container"
    role="presentation"
  >
    <button
      aria-checked="true"
      class="cdr-toggle-button__item"
      role="radio"
      tabindex="0"
      value="thing-2"
    >
      
      thing-2
      
    </button>
  </li>
  <li
    class="cdr-toggle-button__container"
    role="presentation"
  >
    <button
      aria-checked="false"
      class="cdr-toggle-button__item"
      role="radio"
      tabindex="-1"
      value="thing-3"
    >
      
      thing-3
      
    </button>
  </li>
  
</ul>
`;

exports[`CdrToggleButton > component snapshot with custom values > matches the snapshot 1`] = `
<ul
  aria-label="things-food"
  class="cdr-toggle-group cdr-toggle-group--medium"
  role="radiogroup"
>
  
  <li
    class="cdr-toggle-button__container"
    role="presentation"
  >
    <button
      aria-checked="false"
      class="cdr-toggle-button__item"
      role="radio"
      tabindex="-1"
      value="banana,grapes,watermelons"
    >
      
      fruits
      
    </button>
  </li>
  <li
    class="cdr-toggle-button__container"
    role="presentation"
  >
    <button
      aria-checked="false"
      class="cdr-toggle-button__item"
      role="radio"
      tabindex="-1"
      value=""
    >
      
      nofruits
      
    </button>
  </li>
  
</ul>
`;
